File: backend/app/schemas/optimization.py
"""Cost optimization and resource intelligence schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class OptimizationAnalyzeRequest(BaseModel):
    cluster_id: str | None = None
    demo: bool = False
    analysis_window_hours: int = 24
    report_name: str | None = None


class ResourceUtilizationOut(BaseModel):
    id: str
    cluster_id: str | None
    cluster_name: str | None
    namespace: str | None
    resource_type: str
    resource_name: str
    workload_kind: str | None
    cpu_request_millicores: int | None
    memory_request_mb: int | None
    cpu_limit_millicores: int | None
    memory_limit_mb: int | None
    cpu_usage_avg_percent: float | None
    memory_usage_avg_percent: float | None
    cpu_usage_p95_percent: float | None
    memory_usage_p95_percent: float | None
    request_count_avg: float | None
    error_rate_avg_percent: float | None
    restart_count: int
    replicas_desired: int | None
    replicas_ready: int | None
    sample_count: int
    observation_window_hours: int
    monthly_cost_estimate_usd: float | None
    created_at: datetime

    model_config = {"from_attributes": True}


class OptimizationRuleOut(BaseModel):
    id: str
    name: str
    rule_type: str
    description: str
    severity: str
    is_enabled: bool
    parameters: dict[str, Any] | None

    model_config = {"from_attributes": True}


class OptimizationFindingOut(BaseModel):
    id: str
    report_id: str | None
    rule_id: str | None
    utilization_id: str | None
    cluster_id: str | None
    cluster_name: str | None
    namespace: str | None
    resource_type: str
    resource_name: str
    finding_type: str
    severity: str
    title: str
    description: str
    evidence: dict[str, Any] | None
    confidence_score: float | None
    estimated_monthly_savings_usd: float | None
    recommendation: str | None
    remediation: str | None
    ai_explanation: str | None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class CostRecommendationOut(BaseModel):
    id: str
    report_id: str | None
    finding_id: str | None
    cluster_id: str | None
    cluster_name: str | None
    namespace: str | None
    workload_name: str | None
    workload_kind: str | None
    resource_type: str | None
    resource_name: str | None
    optimization_type: str
    status: str
    severity: str
    confidence_score: float | None
    title: str
    description: str | None
    recommendation: str | None
    impact: str | None
    current_cpu_request_millicores: int | None
    current_memory_request_mb: int | None
    current_cpu_usage_avg_percent: float | None
    current_memory_usage_avg_percent: float | None
    current_replicas: int | None
    recommended_cpu_request_millicores: int | None
    recommended_memory_request_mb: int | None
    recommended_replicas: int | None
    estimated_monthly_savings_usd: float | None
    estimated_cpu_savings_cores: float | None
    estimated_memory_savings_gb: float | None
    ai_explanation: str | None
    remediation_yaml: str | None
    evidence: dict[str, Any] | None
    priority: int
    created_at: datetime

    model_config = {"from_attributes": True}


class OptimizationReportOut(BaseModel):
    id: str
    report_name: str
    cluster_id: str | None
    cluster_name: str | None
    status: str
    analysis_window_hours: int
    total_resources_analyzed: int
    total_findings: int
    total_recommendations: int
    estimated_monthly_savings_usd: float
    estimated_annual_savings_usd: float
    optimization_score: float | None
    summary: str | None
    severity_breakdown: dict[str, int] | None
    category_breakdown: dict[str, int] | None
    impacted_resources: list[dict[str, Any]] | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class OptimizationDashboardStats(BaseModel):
    total_recommendations: int
    open_recommendations: int
    implemented_recommendations: int
    in_progress_recommendations: int
    total_findings: int
    critical_findings: int
    high_findings: int
    estimated_monthly_savings_usd: float
    estimated_annual_savings_usd: float
    optimization_score: float
    severity_breakdown: dict[str, int]
    type_breakdown: dict[str, int]
    top_recommendations: list[CostRecommendationOut]


class OptimizationFindingListResponse(BaseModel):
    items: list[OptimizationFindingOut]
    total: int
    page: int
    page_size: int


class CostRecommendationListResponse(BaseModel):
    items: list[CostRecommendationOut]
    total: int
    page: int
    page_size: int


class OptimizationReportListResponse(BaseModel):
    items: list[OptimizationReportOut]
    total: int
    page: int
    page_size: int


class OptimizationAnalysisResponse(BaseModel):
    report: OptimizationReportOut
    findings: list[OptimizationFindingOut]
    recommendations: list[CostRecommendationOut]
    utilization: list[ResourceUtilizationOut]
    stats: OptimizationDashboardStats

File: frontend/src/services/optimization.ts
import { api } from './api'

export interface ResourceUtilization {
  id: string
  cluster_id?: string | null
  cluster_name?: string | null
  namespace?: string | null
  resource_type: string
  resource_name: string
  workload_kind?: string | null
  cpu_request_millicores?: number | null
  memory_request_mb?: number | null
  cpu_usage_avg_percent?: number | null
  memory_usage_avg_percent?: number | null
  cpu_usage_p95_percent?: number | null
  memory_usage_p95_percent?: number | null
  request_count_avg?: number | null
  restart_count: number
  replicas_desired?: number | null
  monthly_cost_estimate_usd?: number | null
}

export interface OptimizationFinding {
  id: string
  report_id?: string | null
  cluster_id?: string | null
  cluster_name?: string | null
  namespace?: string | null
  resource_type: string
  resource_name: string
  finding_type: string
  severity: string
  title: string
  description: string
  confidence_score?: number | null
  estimated_monthly_savings_usd?: number | null
  recommendation?: string | null
  remediation?: string | null
  ai_explanation?: string | null
  status: string
  created_at: string
}

export interface CostRecommendation {
  id: string
  report_id?: string | null
  finding_id?: string | null
  cluster_id?: string | null
  cluster_name?: string | null
  namespace?: string | null
  workload_name?: string | null
  workload_kind?: string | null
  resource_type?: string | null
  resource_name?: string | null
  optimization_type: string
  status: string
  severity: string
  confidence_score?: number | null
  title: string
  description?: string | null
  recommendation?: string | null
  impact?: string | null
  current_cpu_request_millicores?: number | null
  current_memory_request_mb?: number | null
  current_cpu_usage_avg_percent?: number | null
  current_memory_usage_avg_percent?: number | null
  current_replicas?: number | null
  recommended_cpu_request_millicores?: number | null
  recommended_memory_request_mb?: number | null
  recommended_replicas?: number | null
  estimated_monthly_savings_usd?: number | null
  estimated_cpu_savings_cores?: number | null
  estimated_memory_savings_gb?: number | null
  ai_explanation?: string | null
  remediation_yaml?: string | null
  priority: number
  created_at: string
}

export interface OptimizationReport {
  id: string
  report_name: string
  cluster_id?: string | null
  cluster_name?: string | null
  status: string
  analysis_window_hours: number
  total_resources_analyzed: number
  total_findings: number
  total_recommendations: number
  estimated_monthly_savings_usd: number
  estimated_annual_savings_usd: number
  optimization_score?: number | null
  summary?: string | null
  severity_breakdown?: Record<string, number> | null
  category_breakdown?: Record<string, number> | null
  impacted_resources?: Array<Record<string, unknown>> | null
  created_at: string
}

export interface OptimizationStats {
  total_recommendations: number
  open_recommendations: number
  implemented_recommendations: number
  in_progress_recommendations: number
  total_findings: number
  critical_findings: number
  high_findings: number
  estimated_monthly_savings_usd: number
  estimated_annual_savings_usd: number
  optimization_score: number
  severity_breakdown: Record<string, number>
  type_breakdown: Record<string, number>
  top_recommendations: CostRecommendation[]
}

export interface RecommendationListResponse {
  items: CostRecommendation[]
  total: number
  page: number
  page_size: number
}

export interface FindingListResponse {
  items: OptimizationFinding[]
  total: number
  page: number
  page_size: number
}

export interface ReportListResponse {
  items: OptimizationReport[]
  total: number
  page: number
  page_size: number
}

export interface OptimizationAnalysisResponse {
  report: OptimizationReport
  findings: OptimizationFinding[]
  recommendations: CostRecommendation[]
  utilization: ResourceUtilization[]
  stats: OptimizationStats
}

interface ListParams {
  page?: number
  page_size?: number
  cluster_id?: string
  optimization_type?: string
  severity?: string
  status?: string
}

export const optimizationApi = {
  stats: () => api.get<OptimizationStats>('/optimization/stats').then((r) => r.data),

  recommendations: (params?: ListParams) =>
    api.get<RecommendationListResponse>('/optimization/recommendations', { params }).then((r) => r.data),

  recommendation: (id: string) =>
    api.get<CostRecommendation>(`/optimization/recommendations/${id}`).then((r) => r.data),

  findings: (params?: ListParams & { finding_type?: string }) =>
    api.get<FindingListResponse>('/optimization/findings', { params }).then((r) => r.data),

  reports: (params?: ListParams) =>
    api.get<ReportListResponse>('/optimization/reports', { params }).then((r) => r.data),

  analyze: (payload?: { cluster_id?: string; demo?: boolean; analysis_window_hours?: number; report_name?: string }) =>
    api.post<OptimizationAnalysisResponse>('/optimization/analyze', payload ?? { demo: true }).then((r) => r.data),
}
